### PR TITLE
[INTEL MKL] Memory control to avoid caching large buffer for MKL primitives

### PR DESCRIPTION
**Pull request by gzmkl**
This PR enables memory usage control: 
To avoid caching large buffer for MKL primitives, "caching" MKL primitives in the following scenarios is prevented:
1. conv2d backward - always by default (see bottom comment on the environment variable)
2. conv2d forward: large batch size (> 32) and for old systems which do not support 
                               AVX2 or AVX512 (e.g. Ivebridge). 
3. conv2d forward: large batch size (>32) and for 1x1 kernel and stride != 1

Environment variable TF_MKL_OPTIMIZE_PRIMITVE_MEMUSE is "true" by default or not set. 
To enable primitive reuse for all ops and all cases, set it to "false".

### CODE REVIEW

File: tensorflow/core/kernels/mkl_conv_grad_filter_ops.cc [hunk 1]
```diff
@@ -300,19 +300,24 @@ template <typename T>
 class MklConvBwdFilterPrimitiveFactory : public MklPrimitiveFactory<T> {
  public:
   static MklConvBwdFilterPrimitive<T>* Get(
-      const MklConvBwdFilterParams& convBwdFilterDims) {
+      const MklConvBwdFilterParams& convBwdFilterDims, bool not_cache) {
```
Comment (tatianashp) [inline tensorflow/core/kernels/mkl_conv_grad_filter_ops.cc]: Please rename all instances of 'not_cache' into 'do_not_cache'.

Comment (gzmkl) [reply tensorflow/core/kernels/mkl_conv_grad_filter_ops.cc]: Sure

Comment (gzmkl) [reply tensorflow/core/kernels/mkl_conv_grad_filter_ops.cc]: Code change has been done with variable renaming in all occurrences. 
Thanks!

**Review by tatianashp — CHANGES_REQUESTED** [review]
Thanks for improving MKL performance! The code looks good in general. Please address a few minor comments below.

**Comment by tatianashp** [comment]
Could you please also address the comment abou private class member  'bool do_not_cache_'. I don't understand why it's a class member instead of a local variable given that it's used in just one method.

**Comment by gzmkl** [comment]
Sure, I will address your review comments. Thanks

File: tensorflow/core/kernels/mkl_conv_grad_filter_ops.cc [hunk 2]
```diff
@@ -953,6 +966,7 @@ class MklConvCustomBackpropFilterOp
   const int kInputIndex_InputSizes = 0;
   const int kDilationH = 0, kDilationW = 1;
   engine cpu_engine_ = engine(engine::cpu, 0);
+  bool not_cache_;
```
[the remark this answers is not in the document]
Comment (gzmkl) [reply tensorflow/core/kernels/mkl_conv_grad_filter_ops.cc]: Yes, it should be a local variable. Will change.

File: tensorflow/core/kernels/mkl_conv_ops.cc [hunk 3]
```diff
@@ -970,6 +991,7 @@ class MklConvOp : public OpKernel {
   const int kOutputIndex_Dst = 0, kOutputIndex_Filter = 1;
   const int kDilationH = 0, kDilationW = 1;
   engine cpu_engine = engine(engine::cpu, 0);
+  bool not_cache_;
```
[the remark this answers is not in the document]
Comment (gzmkl) [reply tensorflow/core/kernels/mkl_conv_ops.cc]: Yes, I have done the changes with all suggestions and will commit soon after sanity check (build & some validation).

**Comment by gzmkl** [comment]
Hi, I committed in new changes by replacing the private member "do_not_cache_" with a local
variable "do_not_cache", in 3 source files. 

Please let me know if there is any other blocker.

**Review by tatianashp — APPROVED** [review]
Looks good now. Thanks!

**Pull request merged**